{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "81b672a4-f4a2-4d03-a47a-d5f69c0c450e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from typing import List, Dict, Union\n",
    "from teemi.design.combinatorial_design import simple_amplicon_maker\n",
    "from pydna.design import primer_design\n",
    "from pydna.dseqrecord import Dseqrecord\n",
    "\n",
    "\n",
    "def open_gff3_files(path: str = \"\") -> List[List[str]]:\n",
    "    \"\"\"\n",
    "    Opens and reads a GFF3 file and returns its contents as a list of lists.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    path: str\n",
    "        The path to the GFF3 file.\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    List[List[str]]\n",
    "        A list of lists containing the contents of the GFF3 file.\n",
    "    \"\"\"\n",
    "    with open(path, \"r\") as infile:\n",
    "        LINES = []\n",
    "        for line in infile:\n",
    "            LINES.append(line[:].split(\"\\t\"))\n",
    "        LINES = LINES[1:]\n",
    "\n",
    "    return LINES\n",
    "\n",
    "\n",
    "def tidy_up_gff(lst_of_gff: list) -> list:\n",
    "    \"\"\"\n",
    "    This function takes a list of GFF lines and returns a list of dictionaries,\n",
    "    with each dictionary containing information on the signal peptides in the GFF file.\n",
    "\n",
    "    Parameters:\n",
    "    lst_of_gff (list): A list of GFF lines.\n",
    "\n",
    "    Returns:\n",
    "    list_of_peptides (list): A list of dictionaries, with each dictionary containing information on the signal peptides in the GFF file.\n",
    "    \"\"\"\n",
    "    signal_peptides = {}\n",
    "    list_of_peptides = []\n",
    "\n",
    "    for peptide in lst_of_gff:\n",
    "        signal_peptides[\"gene\"] = peptide[0][:19]\n",
    "        signal_peptides[\"start_pos\"] = int(peptide[3]) - 1\n",
    "        signal_peptides[\"end_pos\"] = int(peptide[4]) + 1\n",
    "        signal_peptides[\"signal_peptide_likelyhood\"] = peptide[5]\n",
    "        list_of_peptides.append(signal_peptides)\n",
    "        signal_peptides = {\n",
    "            \"gene\": \"\",\n",
    "            \"start_pos\": \"\",\n",
    "            \"end_pos\": \"\",\n",
    "            \"signal_peptide_likelyhood\": \"\",\n",
    "        }\n",
    "\n",
    "    return list_of_peptides\n",
    "\n",
    "\n",
    "def dict_of_signal_peptides(path: str = \"\") -> List[Dict[str, Union[str, int]]]:\n",
    "    \"\"\"\n",
    "    Given a path to a GFF3 file, returns a list of dictionaries with information on signal peptides.\n",
    "\n",
    "    Args:\n",
    "        path (str): Path to the GFF3 file. Default is an empty string.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of dictionaries where each dictionary contains the following keys:\n",
    "            - 'gene' (str): Gene name of the signal peptide.\n",
    "            - 'start_pos' (int): Start position of the signal peptide in the protein sequence.\n",
    "            - 'end_pos' (int): End position of the signal peptide in the protein sequence.\n",
    "            - 'signal_peptide_likelyhood' (str): The likelihood of the sequence being a signal peptide.\n",
    "    \"\"\"\n",
    "    gff = open_gff3_files(path)\n",
    "    dict_of_signal_peptides = tidy_up_gff(gff)\n",
    "    return dict_of_signal_peptides\n",
    "\n",
    "\n",
    "def read_gff_to_pd(path: str = \"\") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a GFF3 file and returns a pandas DataFrame with columns 'gene', 'start_pos', 'end_pos',\n",
    "    and 'signal_peptide_likelyhood'.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    path : str\n",
    "        The path to the GFF3 file.\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    df : pandas.DataFrame\n",
    "        A DataFrame with columns 'gene', 'start_pos', 'end_pos', and 'signal_peptide_likelyhood'.\n",
    "    \"\"\"\n",
    "\n",
    "    gff = open_gff3_files(path)\n",
    "    dict_of_signal_peptides = tidy_up_gff(gff)\n",
    "    df = pd.DataFrame.from_records(dict_of_signal_peptides)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "import requests\n",
    "import json\n",
    "\n",
    "\n",
    "def primer_ta_neb(primer1, primer2, conc=0.4, prodcode=\"phusion-0\"):\n",
    "    \"\"\"Calculates primer pair melting temp TA,  from NEB.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    primer1 : str\n",
    "        first primer to be used for finding the optimal ta\n",
    "    primer2 : str\n",
    "        second primer to be used for finding the optimal ta\n",
    "    conc : float\n",
    "    prodcode : str\n",
    "        find product codes on nebswebsite: https://tmapi.neb.com/docs/productcodes\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    ta : int\n",
    "        primer pair annealing temp\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    url = \"https://tmapi.neb.com/tm/batch\"\n",
    "    seqpairs = [[primer1, primer2]]\n",
    "\n",
    "    input = {\"seqpairs\": seqpairs, \"conc\": conc, \"prodcode\": prodcode}\n",
    "    headers = {\"content-type\": \"application/json\"}\n",
    "    res = requests.post(url, data=json.dumps(input), headers=headers)\n",
    "\n",
    "    r = json.loads(res.content)\n",
    "\n",
    "    if r[\"success\"]:\n",
    "        for row in r[\"data\"]:\n",
    "            return row[\"ta\"]\n",
    "\n",
    "    else:\n",
    "        print(\"request failed\")\n",
    "        print(r[\"error\"][0])\n",
    "\n",
    "\n",
    "def primer_tm_neb1(primer, conc=0.4, prodcode=\"phusion-0\"):\n",
    "    \"\"\"Calculates a single primers melting temp from NEB.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    primer1 : str\n",
    "    conc : float\n",
    "    prodcode : str\n",
    "        find product codes on nebswebsite: https://tmapi.neb.com/docs/productcodes\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    tm : int\n",
    "        primer melting temperature\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    url = \"https://tmapi.neb.com/tm/batch\"\n",
    "    seqpairs = [[primer]]\n",
    "\n",
    "    input = {\"seqpairs\": seqpairs, \"conc\": conc, \"prodcode\": prodcode}\n",
    "    headers = {\"content-type\": \"application/json\"}\n",
    "    res = requests.post(url, data=json.dumps(input), headers=headers)\n",
    "\n",
    "    r = json.loads(res.content)\n",
    "\n",
    "    if r[\"success\"]:\n",
    "        for row in r[\"data\"]:\n",
    "            return row[\"tm1\"]\n",
    "    else:\n",
    "        print(\"request failed\")\n",
    "        print(r[\"error\"][0])\n",
    "        \n",
    "        \n",
    "def make_amplicons(\n",
    "    list_of_amplicons: list, target_tm=58, limit=10, tm_function=primer_tm_neb1):\n",
    "    \"\"\"Generates pydna.amplicons which contains primers with a target temperature.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    list_of_amplicons : list\n",
    "        list of pydna.Dseqrecords\n",
    "    target_tm : int\n",
    "        representing the target melting temperature for the primers (default=55)\n",
    "    limit: int\n",
    "        representing the minimum primer size (default=5)\n",
    "    tm_function : function\n",
    "        for calculating primer melting temperature (default=primer_tm_neb)\n",
    "\n",
    "    Returns:\n",
    "    amplicons: list\n",
    "        list of amplicon objects with designed primer sequences\n",
    "    \"\"\"\n",
    "    amplicons = []\n",
    "    for i in range(len(list_of_amplicons)):\n",
    "        amplicon = primer_design(\n",
    "            list_of_amplicons[i],\n",
    "            target_tm=target_tm,\n",
    "            limit=limit,\n",
    "            tm_function=tm_function,\n",
    "        )\n",
    "\n",
    "        amplicons.append(amplicon)\n",
    "\n",
    "    return amplicons\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f450746",
   "metadata": {},
   "source": [
    "# Primer design notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f474aec",
   "metadata": {},
   "source": [
    "## Introdution: \n",
    "\n",
    "What you will need: \n",
    "1. Sequences that you want to amplify\n",
    "2. SignalP output for those sequences\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac17ad5c-6122-4ee1-8488-1febe35345de",
   "metadata": {},
   "source": [
    "Steps:\n",
    "\n",
    "1) Read sequences from a text file\n",
    "2) Upload and Process sigP output\n",
    "3) Remove SP from nucleotide sequence\n",
    "4) Make primers\n",
    "\n",
    "\n",
    "\n",
    "Note: if you want to change primer_calculation parameters check the functions above. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e80a60f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "98ed08cb-7618-4f40-b1f1-d14423367b11",
   "metadata": {},
   "source": [
    "## 1. Read sequences from a text file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "dd80c85f-2a8d-46f5-9ca8-a6f2ff27a0a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d059e56c-9d3b-493a-b833-13851ae33437",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: AO090005001389\n",
      "Name: AO090005001389\n",
      "Description: AO090005001389\n",
      "Number of features: 0\n",
      "Seq('ATGGGAGCTCCCCGGCTTCGGATCAAAGGCGCAACCTTCAAAGACCCAAATAAT...TAG')\n",
      "ID: AO090038000444\n",
      "Name: AO090038000444\n",
      "Description: AO090038000444\n",
      "Number of features: 0\n",
      "Seq('ATGCGTTCCTTGTCGTCTATCGCACTCCTGTCTGTGGTGGGAGCTGCGTCTGCC...TAA')\n",
      "ID: AO090012000917\n",
      "Name: AO090012000917\n",
      "Description: AO090012000917\n",
      "Number of features: 0\n",
      "Seq('ATGCGGGCGTCACTACCGTTTCTTACTGCCTTGGGATGTATTCCAGCCGCTTTG...TGA')\n",
      "ID: AO090005001553\n",
      "Name: AO090005001553\n",
      "Description: AO090005001553\n",
      "Number of features: 0\n",
      "Seq('ATGAAATTCCGTAACCTTTTCTTTGCTGCCGTAGCTGGCTCTGCGGTTGCTGCT...TAG')\n",
      "ID: AO090011000715\n",
      "Name: AO090011000715\n",
      "Description: AO090011000715\n",
      "Number of features: 0\n",
      "Seq('ATGAGAATCGGCAACTTGATCGTGGCTGCAAGTGCTGCAAGCCTGGTGCATGCG...TAA')\n",
      "ID: AO090011000757\n",
      "Name: AO090011000757\n",
      "Description: AO090011000757\n",
      "Number of features: 0\n",
      "Seq('ATGAAAGTCACCAGGCTAGCGGTTCTGAATACCCTGGCAACCCTAACTGTTGCC...TAG')\n",
      "ID: AO090003000990\n",
      "Name: AO090003000990\n",
      "Description: AO090003000990\n",
      "Number of features: 0\n",
      "Seq('ATGTTTATTAAACTTTTAAACAAAGCGCTTCTAGTGCTTGGTCTGCTGTCAGCT...TAA')\n",
      "ID: AO090009000373\n",
      "Name: AO090009000373\n",
      "Description: AO090009000373\n",
      "Number of features: 0\n",
      "Seq('ATGACTGACCCAAACTATACACCACCAACCGTGGAGGATATATTTCGCTACCGA...TGA')\n",
      "ID: AO090012000006\n",
      "Name: AO090012000006\n",
      "Description: AO090012000006\n",
      "Number of features: 0\n",
      "Seq('ATGGCAATGCTGTTCACTCCCATTGCGGCTGCATCACTGCTAGCGGTGGTGGGC...TAA')\n",
      "ID: AO090001000604\n",
      "Name: AO090001000604\n",
      "Description: AO090001000604\n",
      "Number of features: 0\n",
      "Seq('ATGCCGAGTCACTCGCGCAGTCGGGACCGCTATCGCAGCGAGCGGGATCCGAGC...TAA')\n",
      "ID: AO090010000122\n",
      "Name: AO090010000122\n",
      "Description: AO090010000122\n",
      "Number of features: 0\n",
      "Seq('ATGGAGCTCTCTTCTGGATCAAGAGCGGAGCATCGCCCGATCCACGCGTGGGCC...TGA')\n",
      "ID: AO090003001590\n",
      "Name: AO090003001590\n",
      "Description: AO090003001590\n",
      "Number of features: 0\n",
      "Seq('ATGCAACCATTACGGCTTTTGGCTCTCACAGCCATTCTTAAGGGCGCCTGGGCC...TGA')\n"
     ]
    }
   ],
   "source": [
    "sequences = list()\n",
    "for sequence in SeqIO.parse('../data/external/GeneByLocusTag.fasta',format='fasta'):\n",
    "    print(sequence)\n",
    "    sequences.append(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0983e267-4ac0-446a-af00-0c29c7afae9d",
   "metadata": {},
   "source": [
    "## 2. Import sigP results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9cdd9952-896d-4e75-879c-f0b86808bdca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>signal_peptide_likelyhood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090038000444</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.9994235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090012000917</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.99983215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005001553</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>0.999823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090011000715</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>0.9998376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090011000757</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.9998413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090003000990</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.9998205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AO090012000006</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0.99984413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AO090003001590</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.99982584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene  start_pos  end_pos signal_peptide_likelyhood\n",
       "0  AO090038000444          0       19                 0.9994235\n",
       "1  AO090012000917          0       20                0.99983215\n",
       "2  AO090005001553          0       28                  0.999823\n",
       "3  AO090011000715          0       28                 0.9998376\n",
       "4  AO090011000757          0       19                 0.9998413\n",
       "5  AO090003000990          0       23                 0.9998205\n",
       "6  AO090012000006          0       27                0.99984413\n",
       "7  AO090003001590          0       19                0.99982584"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "signal_pep = read_gff_to_pd('../data/external/SigP_results/test_3/output.gff3')\n",
    "signal_pep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1de1210d-90c4-478b-989a-302b4e6a0ce0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': ['AO090038000444',\n",
       "  'AO090012000917',\n",
       "  'AO090005001553',\n",
       "  'AO090011000715',\n",
       "  'AO090011000757',\n",
       "  'AO090003000990',\n",
       "  'AO090012000006',\n",
       "  'AO090003001590'],\n",
       " 'n_pos': [19, 20, 28, 28, 19, 23, 27, 19]}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N_pos = signal_pep['end_pos'].to_list()\n",
    "genes_with_signal_peptides = signal_pep['gene'].to_list()\n",
    "\n",
    "my_dict = {'name':genes_with_signal_peptides, 'n_pos':N_pos }\n",
    "my_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87480be8-d1e1-4391-9302-3c71d8a1afe2",
   "metadata": {},
   "source": [
    "# 3. Remove SP from nucleotide sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4814bd0c-f376-46a4-b174-98ade7e838fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_seq = list()\n",
    "for i in range(len(genes_with_signal_peptides)):\n",
    "    for seq in sequences: \n",
    "        if genes_with_signal_peptides[i] == seq.id:\n",
    "            clean_seq.append(seq[N_pos[i]*3:-3])\n",
    "            \n",
    "for seq in sequences:\n",
    "    if seq.id not in genes_with_signal_peptides:\n",
    "        clean_seq.append(seq[3:-3])\n",
    "            \n",
    "len(clean_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5ac043d0-48cb-4a99-ad69-c49466af1213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('GCTGGTCCTTGGGCTCAGTGTGGAGGCAAGTCCTTCTCCGGCTCATCCGAGTGT...AAC'), id='AO090038000444', name='AO090038000444', description='AO090038000444', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CCCCATCCCCGAGTGCAGAGTCCGGAGTATGTCAACTGGACAACTTTCAAAGCC...AAG'), id='AO090012000917', name='AO090012000917', description='AO090012000917', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TCAGTTTTCCAATGTTAGTCTGCCCTTCTTCTTCTTCTTCTTCTTCTTCTTTCT...GGG'), id='AO090005001553', name='AO090005001553', description='AO090005001553', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TCGGGATTCACCTGTACGTCAAGTATTGGCTCTGAAGTTACTGGATCTAACGAA...CTG'), id='AO090011000715', name='AO090011000715', description='AO090011000715', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CTTCCAACAACCGACAAGACAATCACCTCTAGTAATGGAACCGATCTCTTCAAA...AGC'), id='AO090011000757', name='AO090011000757', description='AO090011000757', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ACTATGTCAGTATAACCGTGATGTTACCCCTTTTGCTCTGCATAGTCCCTTACT...TGC'), id='AO090003000990', name='AO090003000990', description='AO090003000990', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GATGCCGAGGGTTGGTACAAGGCTCATCCCGGAATGTCGCGCATCAAGGATGTG...AAA'), id='AO090012000006', name='AO090012000006', description='AO090012000006', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('AGTGCCAATTGTACAGGGTCCTTTGACGCCATCTCGGCCTCTGATTTCGTCGCA...GTT'), id='AO090003001590', name='AO090003001590', description='AO090003001590', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GGAGCTCCCCGGCTTCGGATCAAAGGCGCAACCTTCAAAGACCCAAATAATCGA...ATG'), id='AO090005001389', name='AO090005001389', description='AO090005001389', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ACTGACCCAAACTATACACCACCAACCGTGGAGGATATATTTCGCTACCGATAT...CAT'), id='AO090009000373', name='AO090009000373', description='AO090009000373', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CCGAGTCACTCGCGCAGTCGGGACCGCTATCGCAGCGAGCGGGATCCGAGCAGA...TAT'), id='AO090001000604', name='AO090001000604', description='AO090001000604', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GAGCTCTCTTCTGGATCAAGAGCGGAGCATCGCCCGATCCACGCGTGGGCCCAG...AAG'), id='AO090010000122', name='AO090010000122', description='AO090010000122', dbxrefs=[])]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a36c5615-ea32-4d92-b3bf-012008ca82d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_seq = [Dseqrecord(seq) for seq in clean_seq]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6ff9c275-d5ab-40de-92a3-655bf7a38cd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dseq(-1510)\n",
       "GCTG..GAAC\n",
       "CGAC..CTTG"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_seq[0].seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "516d1c5d",
   "metadata": {},
   "source": [
    "## 4. Make primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1a2a75ec-ba53-47fb-a5ab-b5aa469409a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicons = make_amplicons(clean_seq,\n",
    "                           target_tm=60, # targtet temp\n",
    "                           limit=15,  # min length of primers\n",
    "                           tm_function = primer_tm_neb1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b56a9e5d-c328-43a0-b88f-a513dd3e7098",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCTGGTCCTTGGGCT')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amplicons[0].forward_primer.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "47a38604-c4db-4ca6-96da-e7171db3f678",
   "metadata": {},
   "outputs": [],
   "source": [
    "forward_primer = [str(f.forward_primer.seq) for f in amplicons]\n",
    "r_primer = [str(r.reverse_primer.seq) for r in amplicons]\n",
    "name = [str(r.name) for r in amplicons]\n",
    "gene = [str(r.id) for r in clean_seq]\n",
    "aneal_f = [primer_tm_neb1(str(r)) for r in forward_primer]\n",
    "aneal_r = [primer_tm_neb1(str(r)) for r in r_primer]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "51aa16f9-3fed-4f90-8213-565bff00858b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ta= [primer_ta_neb(str(f.forward_primer.seq),str(f.reverse_primer.seq))  for f in amplicons]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "72011f4a-fbbe-45d5-824e-6e17227d5562",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame({'template': gene, 'f_primer':forward_primer, 'r_primer':r_primer, 'f_tm': aneal_f, 'r_tm': aneal_r, 'ta':ta})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2016bc44-24db-4daa-9117-d220a4760eba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>template</th>\n",
       "      <th>f_primer</th>\n",
       "      <th>r_primer</th>\n",
       "      <th>f_tm</th>\n",
       "      <th>r_tm</th>\n",
       "      <th>ta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090038000444</td>\n",
       "      <td>GCTGGTCCTTGGGCT</td>\n",
       "      <td>GTTCTTGTTGACATTTTCCATATGGTC</td>\n",
       "      <td>57</td>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090012000917</td>\n",
       "      <td>CCCCATCCCCGAGTG</td>\n",
       "      <td>CTTCGTATCATGGAATCTCGTAAGGTC</td>\n",
       "      <td>57</td>\n",
       "      <td>62</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005001553</td>\n",
       "      <td>TCAGTTTTCCAATGTTAGTCTGC</td>\n",
       "      <td>CCCCAGGTATGCCTG</td>\n",
       "      <td>59</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090011000715</td>\n",
       "      <td>TCGGGATTCACCTGTAC</td>\n",
       "      <td>CAGGCACTGGCTGTA</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090011000757</td>\n",
       "      <td>CTTCCAACAACCGACAAGA</td>\n",
       "      <td>GCTGCTTCGACCTTTACA</td>\n",
       "      <td>57</td>\n",
       "      <td>56</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090003000990</td>\n",
       "      <td>ACTATGTCAGTATAACCGTGATG</td>\n",
       "      <td>GCACTGCTTAGGGAACTC</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AO090012000006</td>\n",
       "      <td>GATGCCGAGGGTTGG</td>\n",
       "      <td>TTTACGAGTGATTGTGATAGTGATGT</td>\n",
       "      <td>55</td>\n",
       "      <td>60</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AO090003001590</td>\n",
       "      <td>AGTGCCAATTGTACAGGG</td>\n",
       "      <td>AACAGTCAAAGTAAAGTTGACCG</td>\n",
       "      <td>56</td>\n",
       "      <td>59</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AO090005001389</td>\n",
       "      <td>GGAGCTCCCCGGCTT</td>\n",
       "      <td>CATGATAGTACAGGCTCCACGTTGG</td>\n",
       "      <td>60</td>\n",
       "      <td>64</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AO090009000373</td>\n",
       "      <td>ACTGACCCAAACTATACACC</td>\n",
       "      <td>ATGAACACTGTAAGTGTTGATACTAAC</td>\n",
       "      <td>56</td>\n",
       "      <td>59</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AO090001000604</td>\n",
       "      <td>CCGAGTCACTCGCGC</td>\n",
       "      <td>ATAATATTCCGGTAAATCCCCG</td>\n",
       "      <td>59</td>\n",
       "      <td>57</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AO090010000122</td>\n",
       "      <td>GAGCTCTCTTCTGGATCAAGA</td>\n",
       "      <td>CTTACGACTGTTGATGGCC</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          template                 f_primer                     r_primer  \\\n",
       "0   AO090038000444          GCTGGTCCTTGGGCT  GTTCTTGTTGACATTTTCCATATGGTC   \n",
       "1   AO090012000917          CCCCATCCCCGAGTG  CTTCGTATCATGGAATCTCGTAAGGTC   \n",
       "2   AO090005001553  TCAGTTTTCCAATGTTAGTCTGC              CCCCAGGTATGCCTG   \n",
       "3   AO090011000715        TCGGGATTCACCTGTAC              CAGGCACTGGCTGTA   \n",
       "4   AO090011000757      CTTCCAACAACCGACAAGA           GCTGCTTCGACCTTTACA   \n",
       "5   AO090003000990  ACTATGTCAGTATAACCGTGATG           GCACTGCTTAGGGAACTC   \n",
       "6   AO090012000006          GATGCCGAGGGTTGG   TTTACGAGTGATTGTGATAGTGATGT   \n",
       "7   AO090003001590       AGTGCCAATTGTACAGGG      AACAGTCAAAGTAAAGTTGACCG   \n",
       "8   AO090005001389          GGAGCTCCCCGGCTT    CATGATAGTACAGGCTCCACGTTGG   \n",
       "9   AO090009000373     ACTGACCCAAACTATACACC  ATGAACACTGTAAGTGTTGATACTAAC   \n",
       "10  AO090001000604          CCGAGTCACTCGCGC       ATAATATTCCGGTAAATCCCCG   \n",
       "11  AO090010000122    GAGCTCTCTTCTGGATCAAGA          CTTACGACTGTTGATGGCC   \n",
       "\n",
       "    f_tm  r_tm  ta  \n",
       "0     57    60  61  \n",
       "1     57    62  61  \n",
       "2     59    54  58  \n",
       "3     54    54  58  \n",
       "4     57    56  60  \n",
       "5     57    57  60  \n",
       "6     55    60  59  \n",
       "7     56    59  60  \n",
       "8     60    64  63  \n",
       "9     56    59  60  \n",
       "10    59    57  60  \n",
       "11    58    57  61  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fa58d6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names_f = [name+'_f' for name in gene]\n",
    "gene_names_r = [name+'_r' for name in gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "7c5e8336-e312-4788-b95c-fe75057b5ea6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>concentration</th>\n",
       "      <th>purification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090038000444_f</td>\n",
       "      <td>GCTGGTCCTTGGGCT</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090012000917_f</td>\n",
       "      <td>CCCCATCCCCGAGTG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005001553_f</td>\n",
       "      <td>TCAGTTTTCCAATGTTAGTCTGC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090011000715_f</td>\n",
       "      <td>TCGGGATTCACCTGTAC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090011000757_f</td>\n",
       "      <td>CTTCCAACAACCGACAAGA</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090003000990_f</td>\n",
       "      <td>ACTATGTCAGTATAACCGTGATG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AO090012000006_f</td>\n",
       "      <td>GATGCCGAGGGTTGG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AO090003001590_f</td>\n",
       "      <td>AGTGCCAATTGTACAGGG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AO090005001389_f</td>\n",
       "      <td>GGAGCTCCCCGGCTT</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AO090009000373_f</td>\n",
       "      <td>ACTGACCCAAACTATACACC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AO090001000604_f</td>\n",
       "      <td>CCGAGTCACTCGCGC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AO090010000122_f</td>\n",
       "      <td>GAGCTCTCTTCTGGATCAAGA</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AO090038000444_r</td>\n",
       "      <td>GTTCTTGTTGACATTTTCCATATGGTC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AO090012000917_r</td>\n",
       "      <td>CTTCGTATCATGGAATCTCGTAAGGTC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AO090005001553_r</td>\n",
       "      <td>CCCCAGGTATGCCTG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AO090011000715_r</td>\n",
       "      <td>CAGGCACTGGCTGTA</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AO090011000757_r</td>\n",
       "      <td>GCTGCTTCGACCTTTACA</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AO090003000990_r</td>\n",
       "      <td>GCACTGCTTAGGGAACTC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AO090012000006_r</td>\n",
       "      <td>TTTACGAGTGATTGTGATAGTGATGT</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AO090003001590_r</td>\n",
       "      <td>AACAGTCAAAGTAAAGTTGACCG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AO090005001389_r</td>\n",
       "      <td>CATGATAGTACAGGCTCCACGTTGG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AO090009000373_r</td>\n",
       "      <td>ATGAACACTGTAAGTGTTGATACTAAC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AO090001000604_r</td>\n",
       "      <td>ATAATATTCCGGTAAATCCCCG</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AO090010000122_r</td>\n",
       "      <td>CTTACGACTGTTGATGGCC</td>\n",
       "      <td>25nm</td>\n",
       "      <td>STD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                name                     sequence concentration purification\n",
       "0   AO090038000444_f              GCTGGTCCTTGGGCT          25nm          STD\n",
       "1   AO090012000917_f              CCCCATCCCCGAGTG          25nm          STD\n",
       "2   AO090005001553_f      TCAGTTTTCCAATGTTAGTCTGC          25nm          STD\n",
       "3   AO090011000715_f            TCGGGATTCACCTGTAC          25nm          STD\n",
       "4   AO090011000757_f          CTTCCAACAACCGACAAGA          25nm          STD\n",
       "5   AO090003000990_f      ACTATGTCAGTATAACCGTGATG          25nm          STD\n",
       "6   AO090012000006_f              GATGCCGAGGGTTGG          25nm          STD\n",
       "7   AO090003001590_f           AGTGCCAATTGTACAGGG          25nm          STD\n",
       "8   AO090005001389_f              GGAGCTCCCCGGCTT          25nm          STD\n",
       "9   AO090009000373_f         ACTGACCCAAACTATACACC          25nm          STD\n",
       "10  AO090001000604_f              CCGAGTCACTCGCGC          25nm          STD\n",
       "11  AO090010000122_f        GAGCTCTCTTCTGGATCAAGA          25nm          STD\n",
       "12  AO090038000444_r  GTTCTTGTTGACATTTTCCATATGGTC          25nm          STD\n",
       "13  AO090012000917_r  CTTCGTATCATGGAATCTCGTAAGGTC          25nm          STD\n",
       "14  AO090005001553_r              CCCCAGGTATGCCTG          25nm          STD\n",
       "15  AO090011000715_r              CAGGCACTGGCTGTA          25nm          STD\n",
       "16  AO090011000757_r           GCTGCTTCGACCTTTACA          25nm          STD\n",
       "17  AO090003000990_r           GCACTGCTTAGGGAACTC          25nm          STD\n",
       "18  AO090012000006_r   TTTACGAGTGATTGTGATAGTGATGT          25nm          STD\n",
       "19  AO090003001590_r      AACAGTCAAAGTAAAGTTGACCG          25nm          STD\n",
       "20  AO090005001389_r    CATGATAGTACAGGCTCCACGTTGG          25nm          STD\n",
       "21  AO090009000373_r  ATGAACACTGTAAGTGTTGATACTAAC          25nm          STD\n",
       "22  AO090001000604_r       ATAATATTCCGGTAAATCCCCG          25nm          STD\n",
       "23  AO090010000122_r          CTTACGACTGTTGATGGCC          25nm          STD"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "forward_df = pd.DataFrame({'name': gene_names_f, 'sequence':forward_primer, 'concentration':'25nm', 'purification': 'STD'})\n",
    "reverse_df = pd.DataFrame({'name': gene_names_r, 'sequence':r_primer, 'concentration':'25nm', 'purification': 'STD'})\n",
    "\n",
    "idt_primers_result = pd.concat([forward_df, reverse_df],ignore_index=True)\n",
    "idt_primers_result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95d4a544",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
